Snapshot Testing + Quality Guidelines

jbalsas opened the pull request:
Hey @eduardolundgren, @brunobasto, @mthadley, @bryceosterhaus, I've added some things in this PR worth discussing:
- Moved to Jest for testing so we can easily do snapshot testing of the compiled outputs
- Added prettier to formatting code automatically on `precommit`
- Moved from `jshint` to `eslint-config-liferay`

What do you think?

jbalsas committed:
Uses jest instead of mocha and sinon

jbalsas committed:
Runs jest-codemods

jbalsas committed:
Moves tests

jbalsas committed:
Adds snapshots

jbalsas committed:
Runs prettier manually

jbalsas committed:
Runs prettier automatically on commit

jbalsas committed:
Uses eslint and eslint-config-liferay instead of custom jshint (disables some rules initially)

jbalsas committed:
Fixes lint errors

bryceosterhaus commented on lib/__tests__/index.js:
```diff
@@ -0,0 +1,119 @@
+'use strict';
+
+const assert = require('assert');
```
I think we can get rid of using `assert` and use the assertions that jest provides via [expect](https://facebook.github.io/jest/docs/en/expect.html#content)

bryceosterhaus commented on lib/__tests__/index.js:
```diff
@@ -0,0 +1,119 @@
+const del = require('del');
+const fs = require('fs');
+const metalToolsSoy = require('../../index');
+const sinon = require('sinon');
```
We can also just use the [mocks](https://facebook.github.io/jest/docs/en/mock-function-api.html#content) provided by jest

bryceosterhaus commented on lib/__tests__/index.js:
```diff
@@ -0,0 +1,119 @@
+		restoreStream();
+	});
+
+	it('should compile soy files from/to "src" folder by default', function() {
```
Is there a specific reason to use `it()` over `test()`? Jest uses [test()](https://facebook.github.io/jest/docs/en/api.html#testname-fn) but also has an alias of `it()`

mthadley replied:
I think this is a more of a stylistic choice that jest provides. I don't think they have any plans to deprecate `it`. By providing both, it makes it easier for developers to transition over to `jest` from a larger set of other frameworks.

bryceosterhaus replied:
Thats what I figured. I wasn't sure if there was a "testing philosophy" behind it or test.

bryceosterhaus commented on lib/pipelines/__tests__/compileSoy.js:
```diff
@@ -0,0 +1,411 @@
+			.pipe(compileSoy());
+		stream.on('data', function(file) {
+			assert.strictEqual('simple.soy.js', file.relative);
+			expect(file.contents.toString()).toMatchSnapshot();
```
Rather than always calling `toString()` in `toMatchSnapshot` we can add a [serializer](http://facebook.github.io/jest/docs/en/expect.html#expectaddsnapshotserializerserializer) to make life a bit easier for files

bryceosterhaus reviewed (changes requested):
Added a few comments in regards to the implementation of Jest. I think there are a handful of things we can take advantage of that Jest provides.

I'm all for moving towards Jest. Its been super helpful for us on loop.

jbalsas commented:
Hey @bryceosterhaus, I just tried to do the minimum to get it up and running and was planning to cover those later, but since you mentioned them, I'll update this tomorrow and let you know when it's ready.

Thanks for pointers! ☺️

mthadley reviewed (approved):
I agree with all of the points that @bryceosterhaus brought up. This is probably enough to take advantage of Jest's snapshot features, though.

In the end, `jest` should really be the only testing dependency. We shouldn't need `sinon`, `assert`, `chai`, etc.

jbalsas committed:
Removes assert and sinon dependencies in favour of jest expect and mocks

jbalsas committed:
Adds package-lock

jbalsas committed:
Adds snapshot serializer for templates

jbalsas commented:
Hey @bryceosterhaus, @mthadley, I've pushed some additional commits:

- Removed the `sinon` and `assert` dependencies in favour of jest `expect` and `mocks`
- Added an inline `snapshotSerializer`. Not sure if there's already one out there. If so, please let me know. BTW, @mthadley, what about moving https://github.com/mthadley/metal-jest-serializer to the metal org? 😉
- Added a `package-lock.json` file... yeah, I'm already testing on `npm@5` 😂 

I haven't swtiched to the `test()` syntax since it's not a big gain and as Mike mentions, it might make it harder to switch back and forth from other frameworks.

Please, let me know what you think!

/cc @eduardolundgren, @brunobasto, feel free to add anything as well!

bryceosterhaus commented:
So fancy with your npm5 :sunglasses:

LGTM

brunobasto reviewed (approved):
👍

mthadley commented on package.json:
```diff
@@ -32,6 +39,7 @@
     "gulp-util": "^3.0.7",
     "gulp-wrapper": "^1.0.0",
     "merge": "^1.2.0",
+    "sinon": "^2.3.2",
```
@jbalsas Was this meant to be moved to `dependencies`?

jbalsas replied:
definitely not 🙈

mthadley reviewed (approved):
@jbalsas Nice work, tests look very clean. Also, I will talk to you about moving that package to the metal org.

jbalsas commented:
Merged!! Thanks everyone for the input!!
